Options menu can open in different modes

Satharis opened the pull request:
Changed options so it can be opened with different methods in game vs on the main menu. Currently only displays a setting for tutorials being enabled or disabled in the current game.

As a note: While it technically enables and disables the tutorial correctly, the tutorial code doesn't quite handle the state switching that elegantly. Enabling the tutorials for the first time won't display the initial opening tutorial, and toggling the enabled state off and on seems to make it skip whatever the previous objective was.

Closes #1217

hhyyrylainen commented:
I didn't envision that enabling the tutorial would replay all of the tutorials that were skipped, properly. I think mainly it should work for future tutorials. And the more important way the option is to be used, is to disable the tutorials as there's no way to do that after the initial popup.

hhyyrylainen commented on src/general/OptionsMenu.cs:
```diff
@@ -202,6 +216,12 @@ public class OptionsMenu : Control
+    public enum OptionsMode
+    {
+        MainMenu,
+        MicrobeStage,
```
I think this should be named "InGame"

Satharis replied:
I had it like that because I wasn't sure if the options would need to change when opened from another stage at some point.

hhyyrylainen replied:
I don't foresee there being a need for stage specific options. It will be later confusing when each stage needs to say to the options "yeah, I'm the microbe stage (when I'm really the space stage)"
Even if there are stage specific options, they should all be always visible.

hhyyrylainen commented on src/general/OptionsMenu.cs:
```diff
@@ -258,32 +273,87 @@ public override void _Ready()
         selectedOptionsTab = SelectedOptionsTab.Graphics;
+    }
+
+    public override void _Process(float delta)
```
Empty method

Satharis replied:
It's like that in master, am I supposed to remove it? I think it's only showing as a diff because it got moved.

hhyyrylainen replied:
Please remove it, it unnecessarily runs each frame.

hhyyrylainen commented on src/general/OptionsMenu.cs:
```diff
@@ -314,24 +384,41 @@ public void ApplySettingsToControls(Settings settings)
-    private void SetSettingsTab(string tab)
+    private void SwitchMode(OptionsMode mode)
+    {
+        if (mode == OptionsMode.MainMenu)
```
Switch statement should be preferred when checking enum values

hhyyrylainen commented on src/general/OptionsMenu.cs:
```diff
@@ -314,24 +384,41 @@ public void ApplySettingsToControls(Settings settings)
+            // Show the local game settings if opened from the microbe stage or editor.
+            localSettingsVBox.Show();
+            optionsMode = OptionsMode.MicrobeStage;
+        }
```
Add an error throw if mode is unknown

hhyyrylainen commented on src/general/OptionsMenu.cs:
```diff
@@ -370,12 +457,12 @@ private void SetSettingsTab(string tab)
     private float ConvertSoundBarToDb(float value)
     {
-        return (value - 100) / AUDIO_BAR_SCALE;
+        return GD.Linear2Db(value / 100.0f);
```
Why was this change done? What does this do?

Satharis replied:
It uses a more proper logarithm conversion to decibels, it also handles silencing the audio if it gets down to 0, vs it being at a really low volume like it was before. The mute checkbox still works on top of that. Unless I misunderstood something, that division by 6.0f ends up with notable amounts of error, like 10% volume should be -20 db but it's -15 with that.

hhyyrylainen replied:
I suppose that's fine, then.

hhyyrylainen reviewed (changes requested):
I think you accidentally moved one of the separators:
![2020-09-28_09 53 34 1256](https://user-images.githubusercontent.com/3796411/94402232-75419b00-0174-11eb-924b-ca2ce63d4b8b.png)

this is after:
![2020-09-28_09 53 15 6993](https://user-images.githubusercontent.com/3796411/94402243-783c8b80-0174-11eb-8de0-51f3711bde0a.png)

Also I think that the same checkbox that is used for new tutorial state, should be used for also the in-game tutorial state. I don't particularly like the new section for settings specific to current game, as I can't imagine more settings being put there, definitely no world generation tweaks and such will be put there. Maybe a future difficulty option could be put there. I guess I'll see if GUI people agree with me.

Here's a screenshot of it:
![2020-09-28_10 24 13 9106](https://user-images.githubusercontent.com/3796411/94402472-c782bc00-0174-11eb-9742-0d6bd14d6ed1.png)

Satharis commented:
I didn't accidentally move it, I moved it so it was split into global and current game settings.

I split the tutorials enabled thing into a separate panel because I thought having them be the same button would be incredibly cryptic. You could look at options on the main menu, see it enabled, load a save, open the menu and now it's unchecked. Now you go out to the main menu and it's checked again. I'm not sure what you would write on that button that would easily convey to the user that when on the main menu it is global and applied to all new games, and when in game it only toggles it on that game, and gets saved/loaded with your save.

hhyyrylainen commented:
That's why I had in parenthesis written for (new games), I imagined that it would change to `(for current save / game)` when the options is opened from in-game.

Satharis commented:
I mean I could change it to that but I also thought it was nice that you had the option to enable/disable the global setting from in game as well, with them being distinct options.

hhyyrylainen commented:
I can see both sides, but for logical consistency I'd like to have a separator also before the cheats option.
I'm hoping more people will pitch in on the #gui channel.

hhyyrylainen reviewed (approved):
I think there's nothing left to fix in this one
